Pdb residue string

rob-miller opened the pull request:
<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->
This patch exposes functionality in PDBIO to generate a string for a single residue in PDB or PQR format.  The primary new functionality over the 'selector' mechanism is that residues can be generated with arbitrary chain IDs.  Relevant test code is added.

Less significant features are accessing as a string instead of e.g. using io.StringIO, documenting the preserve_atom_numbering parameter, addressing #1383, and minor cleanup of the PDBIO tests.

This functionality enables generating non-standard 'PDB' files like the attached for investigating residue sidechain interactions.  (suggest 'color by chain' if you load this into a viewer; each 'chain' is a group of potentially interacting PHE residues, with one member of each group superimposed on the N-CA-CB plane.)
[phe-pair.pdb.gz](https://github.com/biopython/biopython/files/6909336/phe-pair.pdb.gz)


With respect to #1383, the discussion appears to have stopped with the question of 'what to do if preserve_atom_numbering=True'.  I submit that this is not the default case, so (1) we should do the correct TER numbering in the default case, (2) for the non-default user this solution is correct for exporting correctly numbered (accounting for TER) input data, and (3) I think this exposes enough functionality for the non-default user to generate their desired output in the remaining cases.

With respect to the PDBIO tests, I believe that the `write_auto_numbering` and `write_preserve_numbering` test labels are currently swapped, and this fixes that as well.

JoaoRodrigues commented:
This makes the code in `PDBIO` a little more convoluted than it needs to be, in my opinion. Could you change the chains of the residues themselves instead before writing? What exactly would be the use case here? Trying to see if the change is worth it.

FWIW, I just created #3679 and was toying with the idea of breaking down the code in more functions, for easier test coverage, so this would play along those lines.

rob-miller commented:
I ran into issues trying to output and color (in a PDB viewer) arbitrary
sets of residues from multiple PDB files, preferably in a simple way to put
in a tutorial.  It seems like the existing selector method requires a bunch
of code to do this and processes the chain for every residue, when what I
really want is to re-use the lines already inside PDBIO.

With the current system to write a single residue I think I need two lines,
one to change the chain ID followed by a PDBIO.save() with a trivial
selector to catch the target residue.  Inside PDBIO.save() every time
though we open/close a file and spin through the chain just to find the
one/few target residues.  For flexibility on the user side I use
io.StringIO to collect the file writes back into strings because as I'm
processing the chain I've not decided that I want to write out this
residue, or maybe I want it after another residue later in the file, or
really I just want to send the data to stdout so I can pipe it into
something else. Doing things correctly I probably want to save/restore the
chain ID back to original state so I can re-use this code later.  After all
this PDB viewer Chimera complains that all my atom serial numbers are
messed up (always start with 1 for every save(), or use
preserve_atom_numbering and that will mess up across structures collected
in same file), and my TER lines don't have any data fields (not that I care
too much).

So the idea was just to pull out the PDBIO residue lines to an accessible
subroutine, but that insidious TER line complicates things because you
can't know when it will be needed.  I tried just storing the data for the
TER string in self, but this was slower in my tests.

I do agree it makes PDBIO more convoluted, but on the user side this way I
Initialize PDBIO's self.atom_number once and then freely mix/match
set_structure()s and residue and/or TER strings as desired - and the serial
numbers are all correct at the end too.

[Just to reiterate from the sample file above, all of this is only useful
for working with structurally aligned pieces so we're not just spitting out
residues randomly all over 3D space.]

Happy to merge in with your error message improvements if you like, and
happy to see it got you apparently catching up on some PDBIO work :-).



On Sun, Aug 1, 2021 at 1:28 AM João Rodrigues ***@***.***>
wrote:

> This makes the code in PDBIO a little more convoluted than it needs to
> be, in my opinion. Could you change the chains of the residues themselves
> instead before writing? What exactly would be the use case here? Trying to
> see if the change is worth it.
>
> FWIW, I just created #3679
> <https://github.com/biopython/biopython/pull/3679> and was toying with
> the idea of breaking down the code in more functions, for easier test
> coverage, so this would play along those lines.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/biopython/biopython/pull/3676#issuecomment-890453796>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZW3QCUKZ76A7VCA5CYPJ3T2TLORANCNFSM5BJAJBCQ>
> .
>

JoaoRodrigues commented:
Catching up indeed :)

The TER statement is something we can try to improve. Some tools write "naked" TER statements (simply TER), while others output a little more info. Most tools I know of don't actually read the info on the TER statement.

I like the idea of refactoring PDBIO to make it a little more modular. That's what I had in mind too when I looked at it yesterday. Would make it much simpler to unit test as well, as opposed to the current style of more like "integration" tests. Having said that, the records are pretty much written atom by atom (line by line), so bundling them would be annoying. We'd have to reverse the parser process in order to be robust, when writing is a much simpler aspect.

Can you post a sample of what you are trying to do? I have the feeling there's a simpler solution here.

rob-miller commented:
On Sun, 1 Aug 2021 at 19:57, João Rodrigues ***@***.***>
wrote:

> Can you post a sample of what you are trying to do? I have the feeling
> there's a simpler solution here.
>
>
Just quickly on this part, have you looked at the pdb file attached to the
first post on this PR?  You might have to go on GitHub, don't think it came
with the emails.

—
> You are receiving this because you authored the thread.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/biopython/biopython/pull/3676#issuecomment-890610612>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZW3QF6T4WFKCFJXT2SI2DT2XNOVANCNFSM5BJAJBCQ>
> .
>

rob-miller commented:
> The TER statement is something we can try to improve. Some tools write "naked" TER statements (simply TER), while others output a little more info. Most tools I know of don't actually read the info on the TER statement.

lennax fixed Biopython's TER records in 7e185d8be91e170d8dd33c19d36fc265b2e9198f, but as you say I'd never noticed the other fields. Nonetheless it does seem like we should do the right thing, and I can't seem to load the redmine.open-bio.org link from that previous commit to see if that explains why it was an issue.

> I like the idea of refactoring PDBIO to make it a little more modular. That's what I had in mind too when I looked at it yesterday. Would make it much simpler to unit test as well, as opposed to the current style of more like "integration" tests. Having said that, the records are pretty much written atom by atom (line by line), so bundling them would be annoying. We'd have to reverse the parser process in order to be robust, when writing is a much simpler aspect.

I do think it is useful to bundle the atoms together as residues, and initially just had this as a convenience function inside my current internal_coords.py -- but assumed this would be criticized for code duplication.  I would *like* to replace the similar _pdb_atom_string in the Biopython internal_coords.py if we changed PDBIO's version to be public.  

> Can you post a sample of what you are trying to do? I have the feeling there's a simpler solution here.

Sorry, just noticed I had re-sequenced parts of my code before generating the previous sample and broke the serial numbers again.  Here's three more views on the same data since I'm playing with them:

[phe-pair-groups.pdb.gz](https://github.com/biopython/biopython/files/6918011/phe-pair-groups.pdb.gz)
[phe-pair-environs.pdb.gz](https://github.com/biopython/biopython/files/6918012/phe-pair-environs.pdb.gz)
[phe-pair-rings.pdb.gz](https://github.com/biopython/biopython/files/6918013/phe-pair-rings.pdb.gz)

Again, they are all intended to be color-by-chain in your structure viewer of choice, and this time have the serial numbers correct throughout.

I don't think you can actually generate these with the current PDBIO short of creating a new (SM)CRA object, and with that you would lose the ability to annotate different residues in the file as these examples do (another advantage of exposing this part of PDBIO as a subroutine).

peterjc commented:
Cross reference PR #989, to fix redmine 2292 https://redmine.open-bio.org/issues/2292 - from my emails:

> Bug #2292: Bio.PDBIO writes TER records without any required fields
> Author: Michal J.
> Status: Resolved
> Bio.PDBIO is happy to write TER records as "TER\n", which is inconsistent with PDB format specification.
> 
> The PDB format requires that TER records have some fields similar to ATOM records:
> '''The TER record has the same residue name, chain identifier, sequence number and insertion code as the terminal residue. The serial number of the TER record is one number greater than the serial number of the ATOM/HETATM preceding the TER.'''
> 
> [See http://www.wwpdb.org/documentation/format23/sect9.html#TER]
> 
> It leads to problem with programs that require correct TER records (like multiple structural alignment program MUSTANG), and crash when they are not found

(with a patch which I think is what Lenna committed)

rob-miller commented:
updated to incorporate #3679

rob-miller commented:
updated to incorporate #3882

@JoaoRodrigues I am still keen to get this in :-)

I want/need the ability to write arbitrary PDB residues with control over chain ID and atom numbering in order to study pairwise residue interactions like:

<img width="881" alt="Screenshot 2022-04-08 at 15 21 56" src="https://user-images.githubusercontent.com/3370432/162511158-30e5c4bc-2d7c-4c4a-a1da-4a829672846d.png">

I support doing this in #3774 and describe it in #3898.  I've implemented this functionality (`pdb_residue_string()`) in the internal_coords module, but really it should be in PDBIO - both logically and to benefit from improvements like #3679 and #3882. 

The implementation suggested here for PDBIO is much better than what I've done in internal_coords, and I would very much like to remove this code from internal_coords in time for a 1.80 release.

Abusing PDB files is an old and venerable technique, and the kind of thing I feel Biopython really should support!

rob-miller commented:
Some explanation on the image above and why it needs this PR:

The image shows 3 pairs of PHE residues from PDB structure 3PBL (Human dopamine D3 receptor), selected as putatively interacting on the basis that their C-alphas are within 6 angstroms of each other (there are much better and more stringent tests in the literature).  The N-Ca-C atoms (hedra) of the first residue in each pair (197-201, 345-346, and 390-394) are aligned using Chi1 transformation matrices exposed by the internal_coords module.

The attached PDB format file presents these pairs for analysis in PDB viewer software, as used to make the screenshot shown.  The key characteristics that I don't think are possible (at least without a lot more work) without this PR are:

* Each pair has its own chain ID (A, B, C)
* Each individual residue has a correctly formatted TER line
* Atom serial numbers are all sequential
* Each residue has a REMARK line annotating description and distance

Clearly the generated PDB file is not a 'structure', and various viewers will handle the issues this addresses differently.  The goal is to generate data that can be interactively selected, inspected and analysed with the minimum of noise - whether that's warnings and complaints or extraneous bonds.

[phe_pairs_4a.pdb.gz](https://github.com/biopython/biopython/files/8620717/phe_pairs_4a.pdb.gz)

rob-miller commented:
My reasons for wanting this merged:

1. It modularizes PDBIO.save(), which is usually considered a good thing, and factors out the residue generation in the same way the ATOM generation is factored out.  There is complication in handling the atom number for the TER string, but it is reasonable to expect to manage this at the chain level and expose this for manually generating PDB files.  Otherwise it is all the same code, so I really thought this would be a trivial change providing additional capability.
2. The current approach of supplying a select() function (a) doesn't support on-the-fly filtering of a chain as easily or efficiently because you have to generate that function and re-scan the chain each time, (b) doesn't support more complicated things like arbitrary chain IDs for coloring in viewers, (c) writes the TER line after each set of residues which may not be desired, (d) writes to a file or stringio which has to be configured separately so less flexible, and (e) probably some other stuff that I've forgotten.
3. It really should not be in the internal coords code, but it is there at least to support the examples in the documentation.  The support for this could and should be removed from internal coords.
4. Everything that writes PDB output should be together in one place so that changes/improvements like PR #3857 get picked up automatically.

peterjc commented:
Could you take a fresh look at this @JoaoRodrigues? Or maybe @etal or @lennax?

JoaoRodrigues commented:
Will do today.

> Message ID: ***@***.***>
>

rob-miller commented:
Update for 1.84
